client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/notices/notices.tsx

```diff
@@ -10,47 +11,91 @@ type Notice = NoticeList.Notice & {
-				() => noticeOperations.removeNotice( lastNotice.id ),
-				NOTICE_TIMEOUT
+interface NoticeOperationsProps {
+	showPatternInsertedNotice: ( pattern: Pattern ) => void;
+	showPatternReplacedNotice: ( pattern: Pattern ) => void;
+	showPatternRemovedNotice: ( pattern: Pattern ) => void;
+}
+
+export interface NoticesProps {
+	noticeOperations: NoticeOperationsProps;
+	noticeUI: ReactNode;
+}
+
+const withNotices = createHigherOrderComponent(
+	< OuterProps, >( InnerComponent: ComponentType< OuterProps > ) => {
+		return ( props: OuterProps & NoticesProps ) => {
+			const [ noticeList, setNoticeList ] = useState< Notice[] >( [] );
+
+			const removeNotice = useCallback(
+				( id: string ) => {
+					setNoticeList( ( current ) => current.filter( ( notice ) => notice.id !== id ) );
+				},
+				[ setNoticeList ]
 			);
-		}
-	}, [ noticeList, noticeOperations ] );
 
-	return <SnackbarList notices={ noticeList } onRemove={ onRemoveNotice } />;
-};
+			const noticeOperations: NoticeOperationsProps = useMemo( () => {
+				const createNotice = ( id: string, content: ReactChild ) => {
+					const existingNoticeWithSameId = noticeList.find( ( notice ) => notice.id === id );
+					if ( existingNoticeWithSameId?.timer ) {
+						clearTimeout( existingNoticeWithSameId.timer );
+						delete existingNoticeWithSameId.timer;
+					}
 
-export const getNoticeContent = ( action: string, pattern: Pattern ) => {
-	const actions: { [ key: string ]: any } = {
-		add: i18n.translate( 'Block pattern "%(patternName)s" inserted.', {
-			args: { patternName: pattern.title },
-		} ),
-		replace: i18n.translate( 'Block pattern "%(patternName)s" replaced.', {
-			args: { patternName: pattern.title },
-		} ),
-		remove: i18n.translate( 'Block pattern "%(patternName)s" removed.', {
-			args: { patternName: pattern.title },
-		} ),
-	};
-
-	return actions[ action ];
-};
+					const newNotice = {
+						id,
+						content,
+						timer: setTimeout( () => {
+							removeNotice( id );
+						}, NOTICE_TIMEOUT ),
+					};
+
+					setNoticeList( ( current ) => [
+						...current.filter( ( notice ) => notice.id !== id ),
+						newNotice,
+					] );
+				};
+
+				return {
+					showPatternInsertedNotice: ( pattern: Pattern ) => {
+						createNotice(
+							'pattern-inserted',
+							i18n.translate( 'Block pattern "%(patternName)s" inserted.', {
+								args: { patternName: pattern.title },
+							} )
+						);
+					},
+					showPatternReplacedNotice: ( pattern: Pattern ) => {
+						createNotice(
+							'pattern-replaced',
+							i18n.translate( 'Block pattern "%(patternName)s" replaced.', {
+								args: { patternName: pattern.title },
+							} )
+						);
+					},
+					showPatternRemovedNotice: ( pattern: Pattern ) => {
+						createNotice(
+							'pattern-removed',
+							i18n.translate( 'Block pattern "%(patternName)s" removed.', {
+								args: { patternName: pattern.title },
+							} )
+						);
+					},
+				};
+			}, [ noticeList, removeNotice ] );
```

Comment:
The memo might not be very useful as the `noticeList` would be updated all the time. Maybe we can remove `useMemo` or use `useRef` to keep the reference of the `noticeList`.

Reply:
Ah, you're right 🤦 I've remove the memoization, thanks 😄